alibi_detect/cd/base.py

```diff
@@ -147,12 +142,15 @@ def preprocess(self, x: Union[np.ndarray, list]) -> Tuple[Union[np.ndarray, list
         -------
         Preprocessed reference data and new instances.
         """
-        if isinstance(self.preprocess_fn, Callable):  # type: ignore[arg-type]
+        if self.preprocess_fn is not None:
             x = self.preprocess_fn(x)
-            x_ref = self.preprocess_fn(self.x_ref) if self.preprocess_at_pred else self.x_ref
-            return x_ref, x
+            if not self.preprocess_at_init and not self.x_ref_preprocessed:
+                x_ref = self.preprocess_fn(self.x_ref)
+            else:
+                x_ref = self.x_ref
+            return x_ref, x  # type: ignore[return-value]
         else:
-            return self.x_ref, x
+            return self.x_ref, x  # type: ignore[return-value]
```

Comment:
This seems to be the same logic repeated for all detectors, perhaps we can pull that out in a follow-up PR?

Similar comment to the pre-process logic at `__init__` as well.

Reply:
Good idea! Issue opened here: https://github.com/SeldonIO/alibi-detect/issues/512